### PR TITLE
Pull request/float to string conversion

### PR DESCRIPTION
**Pull request by romainbrenguier**
This implements the necessary preprocessing and primitive in the string solver to analyse code with `String.valueOf(F)` calls.
The behaviour of the internal function is not guaranteed to perfectly imitate the Java function (this requires more testing), but it works on simple examples.

### CODE REVIEW

File: src/java_bytecode/java_string_library_preprocess.cpp [hunk 1]
```diff
@@ -961,83 +1009,122 @@ codet java_string_library_preprocesst::make_float_to_string_code(
...
 
-  // Subcase of computerized scientific notation
-  case_sci_notation.else_case()=code_assign_function_to_string_expr(
-    string_expr, ID_cprover_string_of_float_func, {arg}, symbol_table);
-  case_list.push_back(case_sci_notation);
+  // For each possible case with have a condition and a string_exprt
+  std::vector<exprt> condition_list;
+  std::vector<string_exprt> string_expr_list;
+
+  // Case of computerized scientific notation
```
Comment (smowton) [inline src/java_bytecode/java_string_library_preprocess.cpp]: Computerized? Hex representation, perhaps?

**Comment by smowton** [comment]
Also should there be test cases included in this PR?

File: src/java_bytecode/java_string_library_preprocess.cpp (hunk 1 above)
Comment (romainbrenguier) [inline src/java_bytecode/java_string_library_preprocess.cpp]: I used the same terminology as the Java documentation https://docs.oracle.com/javase/7/docs/api/java/lang/Float.html#toString(float)
This refers to the case with some number between 1 and 10, a E, and a number representing the exponent of 10. This is still decimal and not hexadecimal.

**Comment by romainbrenguier** [comment]
@smowton good idea, I had prepared some tests for test-gen here https://github.com/diffblue/test-gen/commits/feature/tests-for-string-value-of-float%23271 but adding one of them for CBMC I noticed a potential bug that I now corrected (https://github.com/diffblue/cbmc/pull/1017/commits/2c1211f2bcd100cf0beb4a46fec3af35553b8278)

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+{
+  const floatbv_typet &type=to_floatbv_type(f.type());
+  ieee_float_spect float_spec(type);
+
+  // We will look at the first 5 digits of the fractional part.
+  // shifted is floor(f * 1e5)
+  exprt shifting=constant_float(1e5, float_spec);
+  exprt shifted=round_expr_to_zero(floatbv_mult(f, shifting));
+  // fractional_part is floor(f * 100000) % 100000
+  mod_exprt fractional_part(shifted, from_integer(100000, shifted.type()));
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: `100000` should be put in a constant.

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+    dec_significand, floatbv_of_int_expr(dec_significand_int, float_spec));
+
+  // shifted_float is floor(f * 1e5)
+  exprt shifting=constant_float(1e5, float_spec);
+  exprt shifted_float=
+    round_expr_to_zero(floatbv_mult(fractional_part, shifting));
+
+  // fractional_part_shifted is floor(f * 100000) % 100000
+  mod_exprt fractional_part_shifted(
+    shifted_float, from_integer(100000, shifted_float.type()));
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: use constant for `100000`

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+  // exponent_string = string_of_int(decimal_exponent)
+  string_exprt exponent_string=add_axioms_from_int(
+    decimal_exponent, 3, ref_type);
+
+  // string_expr = concat(string_expr_with_E, exponent_string)
+  return add_axioms_for_concat(string_expr_with_E, exponent_string);
+}
+
+/// add axioms corresponding to the scientific representation of floating point
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: `add` -> `Add`

File: src/solvers/refinement/string_constraint_generator_valueof.cpp [hunk 5]
```diff
@@ -207,11 +89,11 @@ string_exprt string_constraint_generatort::add_axioms_from_bool(
   return res;
 }
 
-/// add axioms to say the string corresponds to the result of String.valueOf(I)
-/// or String.valueOf(J) java functions applied on the integer expression
-/// \par parameters: a signed integer expression, and a maximal size for the
-///   string
-/// representation
+/// Add axioms to say the string corresponds to the result of String.valueOf(I)
```
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_valueof.cpp]: `to say` -> `enforcing` or `stating`

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+/// \param f: a floating point expression
+/// \param spec: specification for floating point
+exprt estimate_decimal_exponent(const exprt &f, const ieee_float_spect &spec)
+{
+  exprt log_10_of_2=constant_float(0.30102999566398119521373889472449302, spec);
+  exprt bin_exp=get_exponent(f, spec);
+  exprt float_bin_exp=floatbv_of_int_expr(bin_exp, spec);
+  exprt dec_exp=floatbv_mult(float_bin_exp, log_10_of_2);
+  binary_relation_exprt negative_exp(dec_exp, ID_lt, constant_float(0.0, spec));
+  // Rounding to zero is not correct for negative number, so we remove 1.
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: If I remember well, when I tried with the correct one, I would get an exception from CBMC because it is not implemented.

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 7]
```diff
@@ -0,0 +1,445 @@
+/*******************************************************************\
+
+Module: Generates string constraints for functions generating strings
+        from other types, in particular int, long, float, double, char, bool
+
+Author: Romain Brenguier, romain.brenguier@diffblue.com
+
+\*******************************************************************/
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: I'm not sure but it seems they are still present on the other CBMC source files

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+#include "string_constraint_generator.h"
+
+
+/// Gets the unbiased exponent in a floating-point bit-vector
+///
+/// TODO: factorize with float_bv.cpp float_utils.h
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \return A 32 bit integer representing the exponent. Note that 32 bits are
+///   sufficient for the exponent even in octuple precision.
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: for strings I usually say `\return a new string expression` so here I could say `\return A new 32 bit integer expression`

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+  return floatbv_typecast_exprt(f, rounding, signedbv_typet(32));
+}
+
+/// Multiplication of expressions representing floating points.
+/// Note that the rounding mode is set to ROUND_TO_EVEN.
+/// \param f: a floating point expression
+/// \param g: a floating point expression
+/// \param rounding: rounding mode
+/// \return An expression representing floating point multiplication.
+exprt floatbv_mult(const exprt &f, const exprt &g)
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: For this particular function it is used in every calls of  add_axioms_from_float_scientific_notation
so regression test for that functionality should also cover that function.

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+/// add axioms corresponding to the integer part of m, in decimal form with no
+/// leading zeroes, followed by '.' ('\u002E'), followed by one or more decimal
+/// digits representing the fractional part of m.
+///
+/// TODO: this specification is not correct for negative numbers and
+/// double precision and floating point value that exceed 100000
+/// \param f: expression representing a float
+/// \param ref_type: refined type for strings
+/// \return a new string expression
+string_exprt string_constraint_generatort::add_axioms_from_float(
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: I'm replacing it with `add_axioms_for_string_of_float`

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+///
+/// A float is represented as $f = m * 2^e$ where $0 <= m < 2$ is the
+/// significand and $-126 <= e <= 127$ is the exponent.
+/// We want an alternate representation by finding $n$ and
+/// $d$ such that $f=n*10^d$. We can estimate $d$ the following way:
+/// $d ~= log_10(f/n) ~= log_10(m) + log_10(2) * e - log_10(n)$
+/// $d = floor(log_10(2) * e)$
+/// Then $n$ can be expressed by the equation:
+/// $log_10(n) = log_10(m) + log_10(2) * e - d$
+/// $n = f /10^d = m * 2^e / 10^d = m * 2^e / 10^(floor(log_10(2) * e))$
```
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: No, once `d` is fixed this is exact, what can happen is that the estimated `d` is to small and in that case we have add 1 to `d` and divide `n` by 10 to get the actual representation.

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 7 above)
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: AFAIK this should be removed, @reuk can confirm

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+/// \return A new 32 bit integer expression representing the exponent.
+///   Note that 32 bits are sufficient for the exponent even
+///   in octuple precision.
+exprt get_exponent(
+  const exprt &src, const ieee_float_spect &spec)
+{
+  exprt exp_bits=extractbits_exprt(
+    src, spec.f+spec.e-1, spec.f, unsignedbv_typet(spec.e));
+
+  // Exponent is in biased from (numbers from -128 to 127 are encoded with
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: typo from -> form

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+  return minus_exprt(
+    typecast_exprt(exp_bits, signedbv_typet(32)),
+    from_integer(spec.bias(), signedbv_typet(32)));
+}
+
+/// Gets the magnitude without hidden bit
+/// \param src: a floating point expression
+/// \param spec: specification for floating points
+/// \return An unsigned value representing the magnitude.
+exprt get_magnitude(const exprt &src, const ieee_float_spect &spec)
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Double-check terminology -- this is getting the mantissa bits, whereas "magnitude" suggests the exponent to me more than the significand.

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+exprt get_significand(
+  const exprt &src, const ieee_float_spect &spec, const typet &type)
+{
+  PRECONDITION(type.id()==ID_unsignedbv);
+  PRECONDITION(to_unsignedbv_type(type).get_width()>spec.f);
+  typecast_exprt magnitude(get_magnitude(src, spec), type);
+  exprt exponent=get_exponent(src, spec);
+  equal_exprt all_zeros(exponent, from_integer(0, exponent.type()));
+  exprt hidden_bit=from_integer((1 << spec.f), type);
+  plus_exprt with_hidden_bit(magnitude, hidden_bit);
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Might be friendlier to use bitwise-or in the circumstance?

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+  // We add axioms:
+  // a1 : 2 <= |res| <= max_size
+  // a2 : starts_with_dot && no_trailing_zero && is_number
+  // starts_with_dot: res[0] = '.'
+  // is_number: forall i:[1, max_size[. '0' < res[i] < '9'
+  // no_trailing_zero: forall j:[2, max_size[. !(|res| = j+1 && res[j]='0')
+  // a3 : int_expr = sum_j 10^j (j < |res| ? res[j] - '0' : 0)
+
+  and_exprt a1(res.axiom_for_is_strictly_longer_than(1),
+               res.axiom_for_is_shorter_than(max));
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Possibly in a different PR: "is shorter than" means <, not <=. I see the actual function does implement <= however. Suggest renaming `axiom_for_shorter_than` and `axiom_for_shorter_or_equal` and similarly >.

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+  // `dec_exponent` is $d$ in the formulas
+  // it is obtained from the approximation, checking whether it is not too small
+  binary_relation_exprt estimate_too_small(
+    dec_significand_int, ID_ge, from_integer(10, int_type));
+  if_exprt decimal_exponent(
+    estimate_too_small,
+    plus_exprt(dec_exponent_estimate, from_integer(1, int_type)),
+    dec_exponent_estimate);
+
+  // `dec_significand` is divided by 10 if it exeeds 10
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: typo

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+{
+  return add_axioms_for_string_of_float(
+    args(f, 1)[0], to_refined_string_type(f.type()));
+}
+
+/// Add axioms corresponding to the integer part of m, in decimal form with no
+/// leading zeroes, followed by '.' ('\u002E'), followed by one or more decimal
+/// digits representing the fractional part of m.
+///
+/// TODO: this specification is not correct for negative numbers and
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Check is this still accurate? I note the function below suggests for  numbers > 100000 we use scientific notation; does that mean this version will never be used for such numbers or do we still have  a shortcoming here?

File: src/solvers/refinement/string_constraint_generator_float.cpp [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+  // $m$ from the formula is a value between 0.0 and 2.0 represented
+  // with values in the range 0x000000 0xFFFFFF so 1 corresponds to 0x800000.
+  // `bin_significand_int` represents $m * 0x800000$
+  exprt bin_significand_int=
+    get_significand(f, float_spec, unsignedbv_typet(32));
+  // `bin_significand` represents $m$ and is obtained
+  // by multiplying `binary_significand_as_int` by 1/0x800000=1.192092896e-7.
+  exprt bin_significand=floatbv_mult(
+    floatbv_typecast_exprt(bin_significand_int, round_to_zero_expr, float_type),
+    constant_float(1.192092896e-7, float_spec));
```
Comment (smowton) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Not an expert, but check if there's a cleaner way to represent this and the constant tables below?

**Comment by reuk** [comment]
Re: file headers, the rule is to use the `/*****\` style in CBMC and the simpler `// Copyright...` style in private repos.

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 15 above)
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: There is already an issue for that: https://github.com/diffblue/test-gen/issues/321.
It should be addressed at some point in a separate PR.

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 17 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: We can't know when adding the axioms whether the number is greater than 10000, the axioms we add ensure that if the input is smaller than 10000 then the output is correct.
But maybe the TODO is misleading here, it should rather be the specification of the function.

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 13 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: other part of CBMC use `fraction` so I will change for that

File: src/solvers/refinement/string_constraint_generator_float.cpp (hunk 18 above)
Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: I will replace it with hexadecimal `0x2p-6` which is more precise

Comment (romainbrenguier) [inline src/solvers/refinement/string_constraint_generator_float.cpp]: Actually not all compilers support hexadecimal floating points, so I will have to revert to the non-hexadecimal representation.
For the table, a cleaner way to give it would be to recompute it each time, but since it involves relatively complicated operations and they are constants I'm not sure it's better.

**Comment by romainbrenguier** [comment]
@smowton 

1. I think I answered all comments now
2. the problem with adding more tests is that they can be quite slow, each of them could take up to 10s,
    I have some that I plan to add to test-gen and mark as `THOROUGH`

**Comment by romainbrenguier** [comment]
@smowton I added some more tests with different cases. Some pass and are marked as THOROUGH (one takes more than 1minute to complete) and some seem to fail so I marked them as KNOWNBUG, and I will raise separate issues to investigate that.
